### ISSUE TITLE
API shape

### ISSUE THREAD

**Issue by neilalexander** [tetratelabs/wazero#425]
I've been trying to pick up all of the recent changes since changing from `gasm` to `wazero` and I am really struggling with the new API shape. There's obviously a high rate of change, which is nice, but many of the examples in the comments are now obsolete and even code samples from GitHub issues from a couple weeks ago are no longer relevant.

Not to mention that suddenly the `wasm` and `wasi` packages — which I was happily using before — are now suddenly hidden away in `internal` and therefore private, so basically the entire API I was used to using in `gasm` is now seemingly gone. *Ouch*.

Let's say I want to load a WASI application. I create a runtime called `runtime`. I then need `wazero.WASISnapshotPreview1()`, which is a `*wazero.Module`, but `runtime.InstantiateModule` expects a `wasm.Module`. So I use `runtime.InstantiateModule`, which gives me a `wasm.Module`. At this point it isn't clear what the difference is between `wasm.Module` and `*wazero.Module`.

I then want to load my own functions into `env`, so I create a `runtime.NewModuleBuilder`, export some functions and then `Instantiate`, which should instantiate the module back into `runtime` and then returns a `wasm.Module` too. 

Finally, I want to load my own application: `runtime.CompileModule` which returns a `*wazero.Module`, and then I pass that into `runtime.InstantiateModule` which returns a `wasm.Module`.

So I've now got a `wazero.Runtime` and three `wasm.Module`s: the WASI bit, `env` and my application module. 

Now I have a problem. `wazero.StartWASICommandWithConfig` doesn't want a `wasm.Module`, it wants a `*wazero.Module`, and `wazero.StartWASICommandWithConfig` is the only way to pass in a `*wazero.SysConfig` that lets me define stdout, stdin etc. The alternative seemingly is that I can use `module.ExportedFunction("_start").Call`, but that doesn't take a config that lets me define stdout, stdin etc.

If I want to get an exported function or variable from my application module later, well, that only works if it's a `wasm.Module` and not a `*wazero.Module` — the former has `ExportedFunction` whereas the latter doesn't. So it's once again not clear what I'm supposed to do here and why there are two types for seemingly the same concept?

I suppose this is part frustration and part cry for help — sorry about that! Frustration because the `wasm` and `wasi` packages that I had access to made a lot more sense and gave me a lot more control over what was going on, and cry for help because now the new API is all over the place with a heap of confusingly named and yet seemingly incompatible types and the pieces don't seem to fit together. The documentation and comments don't seem to be enlightening either.

Advice appreciated on if I'm missing something obvious.

**Comment by codefromthecrypt** [comment]
Vent warranted and expected under thrash, we have put a goal of API stability in the next month or two so it is a good time to get feedback on things we can change without breaking functionality others need.

### Skip this part if you want to get to the answers

FWIW the way we've been addressing thrash has been in a fews ways:
1. pinging users we know to be active (sorry missed you)
2. Adding godoc to every function and symbol exported
3. raising drift update PRs on every [public project](https://github.com/tetratelabs/wazero/network/dependents that consumes wazero, unless they say stop.
4. ensuring if our examples don't compile, it breaks the build https://github.com/tetratelabs/wazero/tree/main/examples
5. having a RATIONALE document that keeps track of main ideas https://github.com/tetratelabs/wazero/blob/main/RATIONALE.md

We hadn't done some things we could have (plus other things I'm unaware of)
1. Make a web site
2. go back through GH issues and look for any code snippets and update them
3. Handle drift automatically within our godoc (#426)
4. make a migration guide for the "gasm" api (this won't happen as we can do it one-off)

Also, there are still things we just haven't finished yet, so ex there's one project we can't update until that happens.

Many of gaps are due to time, as takeshi and I are already working 10-12 hour days and losing weekends to get the infrastructure stable for 1.0 API. I personally promise to do a web site once it finalizes though.

Anyway, I see what you mean on the godoc, they drifted and that's super annoying. We probably could be better at advertising examples.It was a non-goal to keep the gasm api as that was not a goroutine-safe API and not crafted in a way to become one. The side effect of changing that entirely was an issue like this. OTOH, we do have folks using the current API, so I think we can bridge the gap with you and ideally improve it along the way.

### The answers

Apologies for the misses and I'll try to help. Let's use one example to help with this and forget about the gasm API, let's focus on the API we have which is very much like what 1.0 will be: https://github.com/tetratelabs/wazero/blob/main/examples/file_system_test.go

First thing I notice in your notes is we can do better instructing the dependency tree. 

#### Adding module dependencies you need to the runtime
It is confusing without knowing WebAssembly, that WASI is a dependency (something you import). Similar to a line in your go.mod file, you can't use WASI, or start a WASI command, unless it has been added to the runtime.

So, Instantiate (a term from WASM) is the only way to add a module to the runtime. Here is where we add WASI under its default module name. "wasi_snapshot_preview1"

```
r := wazero.NewRuntime()
// Instantiate WASI, which implements system I/O such as console output.
wasi, _ := r.InstantiateModule(wazero.WASISnapshotPreview1())
defer wasi.Close()
```

Let's say you need to import "env" for AssemblyScript abort handler. Again, we need to add that to the runtime before we can use it.

```
abort := func(m wasm.Module, messageOffset, fileNameOffset, line, col uint32) {
	_ = m.CloseWithExitCode(255)
}
env, _ := r.NewModuleBuilder("env").ExportFunction("abort", abort).Instantiate()
defer env.Close()
```

#### starting your code
Now, you have the dependencies you need (ex things in go.mod needed for import to work), and you can define and run your WASM module.

Since there's currently only one API that can do I/O, the configuration for it is currently pinned to StartWASIXXX. These run the "_start" function by default as that's what the spec says starting one should do.
```
	// Users often have to run their code multiple times, so this "compile" step exists. Maybe we
	// should rename the result to "CompiledModule"? Those running multiple times concurrently
	// do compiled.WithName(next) to avoid conflicts!
	compiled, _ := r.CompileModule(yourWasm)

	// Configure what this module will use in its sandbox
	sysConfig := wazero.NewSysConfig().WithStdout(os.Stdin).

	myModule, _ := wazero.StartWASICommandWithConfig(r, compiled, sysConfig.WithArgs("my-prog", "1", "2"))
	defer myModule.Close()
```

Use whatever you need from myModule. The `wasm.Module` only allows access to exported types, just like go. You can't use an internal function except inside wasm.

We are trying to balance the needs of those who need concurrency and prefer init failures (ex compile before instantiate) with first time usage (both at the same time ex. WithSource). 

I'll leave this here so you can process it before we take any next steps. If you put a demo project up I can try to help keep things current for you as well!

**Comment by codefromthecrypt** [comment]
ps since (my) talk is cheap, and we are in different timezones, here are a few more places that may help show not tell:

https://github.com/tetratelabs/wazero/tree/main/examples
https://github.com/birros/wazero-demo
https://github.com/wapc/wapc-go/blob/master/engines/wazero/wazero.go

I'll watch for your feedback, and you can close this issue when sufficient actionable issues are raised.

**Comment by codefromthecrypt** [comment]
was talking to @mathetake offline and we also didn't like using the word module twice and prefixing it isn't much better. that module is used in different places makes more sense in go (ex go.mod current module vs required ones) as the tooling is implicit. Doing that here is more harm than good. We'll choose another name and ping you (and others active) on the PR.

**Comment by neilalexander** [comment]
First of all I'd just like to say thank you so much for taking this on so quickly. I've been reading through the changes in #427  and wapc/wapc-go#21 and it's undoubtedly much clearer now with the improved naming — and the addition of `InstantiateModuleWithConfig` has pretty much solved my issue of not being able to set stdout and the environment. 

I'll be spending some more time over the weekend on this so I'll try to spend some time looking through the documentation as well to see if there is anything else that doesn't make sense.

Next up on the list is trying to handle string and multiple return values.

**Comment by codefromthecrypt** [comment]
Here's an example of multiple return values https://github.com/tetratelabs/wazero/blob/main/examples/results_test.go

I'll work on one for errors next

**Comment by codefromthecrypt** [comment]
re-org'd the examples dir https://github.com/tetratelabs/wazero/tree/main/examples

**Comment by codefromthecrypt** [comment]
another example https://github.com/tetratelabs/wazero/pull/466

**Comment by codefromthecrypt** [comment]
examples for memory allocation https://github.com/tetratelabs/wazero/pull/475

**Comment by codefromthecrypt** [comment]
OK the examples folder has both multiple return values and strings now. If you are ok, let's close this as issues are better when they can pinpoint a concrete exit!

**Comment by codefromthecrypt** [comment]
we're earnestly trying to stabilize the API without cementing design flaws. this is an example where we had something not good and fixed it. Per README we are trying to get dents knocked out shortly ideally in the next month

https://github.com/tetratelabs/wazero/issues/516#issuecomment-1114464031

**Comment by codefromthecrypt** [comment]
website is up https://wazero.io/

I'll move the docs we have, and if I don't hear back about anything in the next week, I'll close this out vs spamming you!

**Comment by neilalexander** [comment]
The documentation and API shape improvements are hugely welcomed, thanks so much for putting the time and effort so quickly into it! I'll close this issue since from my perspective it is now much clearer how to get going with the latest versions of the library and in particular the allocation examples for TinyGo and Rust are great.
